### PR TITLE
Rack_codebreaker

### PR DESCRIPTION
**Pull request by luftkluft**
Web application.

### CODE REVIEW

File: lib/codebreaker.rb [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+    def validated_data
+      { player_name: @request.session[:player_name] }
+    end
+
+    def update_game
+      @player_name = @request.session[:player_name]
+      @level = @request.session[:level]
+      @attempts_count = @request.session[:attempts_counter]
+      @hints_count = if @request.session[:hints_counter].negative?
```
Comment (dzemlianoi-double) [inline lib/codebreaker.rb]: try to rethink

File: lib/views/rules.html.erb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+</head>
+<body>
+  <section id="introduction">
+    <div class="container">
+      <% @read_rules.each do |line|%>
+      <h5 class="text-white title"><%= line %></h5>
+      <% end %>
+    </div>
+    <div class="errors-info">
+		   <a href='/'>Go to menu</a>
```
Comment (dzemlianoi-double) [inline lib/views/rules.html.erb]: try to use formatter for your html

File: lib/views/win.html.erb [hunk 3]
```diff
@@ -0,0 +1,66 @@
+<!DOCTYPE html>
+<html lang="en" class="mdl-js">
+<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+<link href="assets/favicon.ico" rel="icon">
+<link href="assets/css/bootstrap.min.css" rel="stylesheet">
+<link href="assets/css/styles.css" rel="stylesheet">
+<script src="assets/js/jquery.min.js"></script>
```
Comment (dzemlianoi-double) [inline lib/views/win.html.erb]: move to yarn/npm

File: spec/codebreaker/racker_spec.rb [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+      it 'contains a greeting' do
+        expect(last_response.body).to include('Hello, Name!')
+      end
+    end
+  end
+
+  describe 'Answer submit button click' do
+    it 'click with number `1234` with decrease attempts counter' do
+      get '/'
+      post '/submit_menu_button', player_name: 'Name', level: 'hard'
```
Comment (dzemlianoi-double) [inline spec/codebreaker/racker_spec.rb]: do not walk around many endpoints in one spec

File: config.ru [hunk 5]
```diff
@@ -0,0 +1,9 @@
+require './lib/codebreaker'
+
+app = Rack::Builder.new do
```
[the remark this answers is not in the document]
Comment (luftkluft) [reply config.ru]: Changed entry point.

File: lib/codebreaker.rb [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+  WIN = '++++'.freeze
+  LOSE = 'lose'.freeze
+  SCORE_DATABASE = './lib/data/score.yml'.freeze
+  class Racker
+    def self.call(env)
+      new(env).response.finish
+    end
+
+    def initialize(env)
+      File.open(SCORE_DATABASE, 'a') { |f| f.write([].to_yaml) } unless File.exist?(SCORE_DATABASE)
```
[the remark this answers is not in the document]
Comment (luftkluft) [reply lib/codebreaker.rb]: Moved to '/statistics'

File: lib/codebreaker.rb [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    end
+
+    def setup_player_session
+      @request.session[:player_name] = @request.params['player_name']
+      @request.session[:level] = @request.params['level']
+    end
+
+    def menu_render
+      check_result = @game_init.check_game_data(validated_data)
+      return update_game if check_result == ''
```
[the remark this answers is not in the document]
Comment (luftkluft) [reply lib/codebreaker.rb]: Ok. It is worked.

File: spec/codebreaker/racker_spec.rb (hunk 4 above)
Comment (luftkluft) [reply spec/codebreaker/racker_spec.rb]: Test variable moved to constants.

File: lib/autoload.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+WIN = '++++'.freeze
```
Comment (dzemlianoi-double) [inline lib/autoload.rb]: why this constants is here? remove it or move it to another place

File: lib/codebreaker.rb [hunk 9]
```diff
@@ -0,0 +1,240 @@
+require_relative 'autoload'
+
+module Codebreaker
+  class Racker
+    def self.call(env)
+      new(env).response.finish
+    end
+
+    def initialize(env)
+      @answer = ''
```
Comment (dzemlianoi-double) [inline lib/codebreaker.rb]: do you need @answer at **initialize**?
or only when we call it in some method?

Check another variables

File: lib/codebreaker.rb [hunk 10]
```diff
@@ -0,0 +1,240 @@
+require_relative 'autoload'
```
Comment (dzemlianoi-double) [inline lib/codebreaker.rb]: your file is lib/codebreaker.rb - here must be just codebreaker module.

You mustn't declare include another classes here

File: lib/codebreaker.rb [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+      when '/statistics' then statistics
+      else
+        error404
+      end
+    end
+
+    private
+
+    def index
+      return Rack::Response.new(render('menu.html.erb')) unless File.exist?(HISTORY_DATABASE)
```
Comment (dzemlianoi-double) [inline lib/codebreaker.rb]: you mustn't check History here.
if you have an error - it's some misslogic in your application. try to rethink

File: lib/codebreaker.rb [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+
+      read_history
+    end
+
+    def submit_menu_button
+      unless File.exist?(HISTORY_DATABASE)
```
Comment (dzemlianoi-double) [inline lib/codebreaker.rb]: the same

File: lib/codebreaker.rb [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+
+    private
+
+    def index
+      return Rack::Response.new(render('menu.html.erb')) unless File.exist?(HISTORY_DATABASE)
+
+      read_history
+    end
+
+    def submit_menu_button
```
Comment (dzemlianoi-double) [inline lib/codebreaker.rb]: your method naming is very strange?
where is button? what do we submit?

File: lib/codebreaker.rb [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+        hints: @request.session[:hints],
+        secret_code: @request.session[:secret_code],
+        date: Time.now.strftime('%d-%m-%Y %R') }
+    end
+
+    def write_history
+      File.open(HISTORY_DATABASE, 'w') { |f| f.write(preparation_hystory.to_yaml) }
+    end
+
+    def read_history
```
Comment (dzemlianoi-double) [inline lib/codebreaker.rb]: i don't know what to say about it...

File: lib/codebreaker.rb (hunk 10 above)
Comment (luftkluft) [reply lib/codebreaker.rb]: Autoload removed. Constants transferred to the library. Now only the library is connected.

File: lib/autoload.rb (hunk 8 above)
Comment (luftkluft) [reply lib/autoload.rb]: Removed.

File: lib/codebreaker.rb (hunk 9 above)
Comment (luftkluft) [reply lib/codebreaker.rb]: Unnecessary variables removed.

**Pull request merged**